### PR TITLE
Fix numpy expand and make numpy backend jittable

### PR DESCRIPTION
**Pull request by fehiepsi**
Following fixes are detected while I did profiling GaussianHMM in #315.
- Support `-1` in expand shape for numpy backend using @fritzo 's suggestion
- `UnshapedArray` is replaced by `jax.core.Tracer`. A device array `x` will become `Tracer` under jit/vmap/pmap. This tracer has an attribute `x.aval`, which is UnshapedArray.
- Make `eager_cat_homogeneous` backend agnostic

Follow-up PR:
- enhance the speed of Gaussian

### CODE REVIEW

File: funsor/tensor.py [hunk 1]
```diff
@@ -326,7 +326,7 @@ def unscaled_sample(self, sampled_vars, sample_inputs):
         #   Then g is an unbiased estimator of f in value and all derivatives.
         #   In the special case f = detach(f), we can simplify to
         #       g = delta(x=x0) |f|.
-        if flat_logits.requires_grad:
+        if hasattr(flat_logits, "requires_grad") and flat_logits.requires_grad:
```
Comment (fehiepsi) [inline funsor/tensor.py]: I am not sure if for numpy backend, we should exercise `if` or `else` branch here.

Comment (fritzo) [inline funsor/tensor.py]: 1. Does JAX have a `.detach()` method, as needed in the DiCE branch? Does jax have a `.requires_grad` checker?
2. Should we have three backends, "numpy", "jax", and "torch"? In that case, "numpy" should use the second branch and jax should use the first (i.e. DiCE) branch.

Comment (eb8680) [reply funsor/tensor.py]: > Should we have three backends, "numpy", "jax", and "torch"?

+1 to this, with "numpy" as the default backend - that way dependencies heavier than numpy are optional and backend-specific

Comment (neerajprad) [reply funsor/tensor.py]: > Does JAX have a .detach() method, as needed in the DiCE branch? Does jax have a .requires_grad checker?

The analog of `detach` is [lax.stop_gradient](https://jax.readthedocs.io/en/latest/_autosummary/jax.lax.stop_gradient.html#jax.lax.stop_gradient), but I don't think there is a way to check for `requires_grad` flag (and probably wouldn't work under jit anyways). I suspect that we may have to write this as a JAX custom primitive and specify the behavior under ad.

Comment (fehiepsi) [reply funsor/tensor.py]: Thank all! I can understand the situation now. I'll address this in an upcoming PR, where I add global BACKEND "numpy"/"torch"/"jax". As @neerajprad pointed out, we can use `lax.stop_gradient` in this branch.

About `requires_grad`, I think we don't need to worry about it. If JAX users don't want to take `grad` here, the overhead of `if` branch is pretty small. Please let me know if I miss crucial points here (e.g. `sb_inputs` is different from `batch_inputs`...)

Comment (fritzo) [reply funsor/tensor.py]: OK, so as I understand, in this branch you will
```py
if backend == "torch" and not flat_logits.requires_grad:
    # use the shorcut
else:  # numpy, jax, and torch with grad
    # use the DiCE version
```
and in a follow-up PR you can add "numpy"/"jax"/"torch" and optimize the numpy branch?
```py
if backend == "numpy" or (backend == "torch" and not flat_logits.requires_grad):
    # use the shorcut
else:  # jax, and torch with grad
    # use the DiCE version
```

Comment (fehiepsi) [reply funsor/tensor.py]: Yes, but let's keep this PR as-is because the logic here (for backends other than `torch`) is not exercised in any test yet.
